// @solana-mobile/dapp-store-cli@0.7.1/lib/commands/ValidateCommand.js
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _ts_generator(thisArg, body) {
    var f, y, t, g, _ = {
        label: 0,
        sent: function() {
            if (t[0] & 1) throw t[1];
            return t[1];
        },
        trys: [],
        ops: []
    };
    return(g = {
        next: verb(0),
        "throw": verb(1),
        "return": verb(2)
    }, typeof Symbol === "function" && (g[Symbol.iterator] = function() {
        return this;
    }), g);
    function verb(n) {
        return function(v) {
            return step([
                n,
                v
            ]);
        };
    }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while(_)try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [
                op[0] & 2,
                t.value
            ];
            switch(op[0]){
                case 0:
                case 1:
                    t = op;
                    break;
                case 4:
                    _.label++;
                    return {
                        value: op[1],
                        done: false
                    };
                case 5:
                    _.label++;
                    y = op[1];
                    op = [
                        0
                    ];
                    continue;
                case 7:
                    op = _.ops.pop();
                    _.trys.pop();
                    continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                        _ = 0;
                        continue;
                    }
                    if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
                        _.label = op[1];
                        break;
                    }
                    if (op[0] === 6 && _.label < t[1]) {
                        _.label = t[1];
                        t = op;
                        break;
                    }
                    if (t && _.label < t[2]) {
                        _.label = t[2];
                        _.ops.push(op);
                        break;
                    }
                    if (t[2]) _.ops.pop();
                    _.trys.pop();
                    continue;
            }
            op = body.call(thisArg, _);
        } catch (e) {
            op = [
                6,
                e
            ];
            y = 0;
        } finally{
            f = t = 0;
        }
        if (op[0] & 5) throw op[1];
        return {
            value: op[0] ? op[1] : void 0,
            done: true
        };
    }
}
import { createAppJson, createPublisherJson, createReleaseJson, validateApp, validatePublisher, validateRelease, metaplexFileReplacer } from "@solana-mobile/dapp-store-publishing-tools";
import { debug } from "../CliUtils.js";
import { loadPublishDetailsWithChecks } from "../config/PublishDetails.js";
export var validateCommand = function() {
    var _ref = _async_to_generator(function(param) {
        var signer, buildToolsPath, _ref, publisherDetails, appDetails, releaseDetails, publisherJson, _publisherJson_image, appJson, _appJson_image, releaseJson, objStringified;
        return _ts_generator(this, function(_state) {
            switch(_state.label){
                case 0:
                    signer = param.signer, buildToolsPath = param.buildToolsPath;
                    return [
                        4,
                        loadPublishDetailsWithChecks(buildToolsPath)
                    ];
                case 1:
                    _ref = _state.sent(), publisherDetails = _ref.publisher, appDetails = _ref.app, releaseDetails = _ref.release;
                    debug({
                        publisherDetails: publisherDetails,
                        appDetails: appDetails,
                        releaseDetails: releaseDetails
                    });
                    publisherJson = createPublisherJson(publisherDetails);
                    if (typeof publisherJson.image !== "string") {
                        ;
                        publisherJson.image = (_publisherJson_image = publisherJson.image) === null || _publisherJson_image === void 0 ? void 0 : _publisherJson_image.fileName;
                    }
                    debug("publisherJson=", JSON.stringify({
                        publisherJson: publisherJson
                    }, metaplexFileReplacer, 2));
                    try {
                        validatePublisher(publisherJson);
                        console.info("Publisher JSON valid!");
                    } catch (e) {
                        console.error(e);
                    }
                    appJson = createAppJson(appDetails, signer.publicKey);
                    if (typeof appJson.image !== "string") {
                        ;
                        appJson.image = (_appJson_image = appJson.image) === null || _appJson_image === void 0 ? void 0 : _appJson_image.fileName;
                    }
                    debug("appJson=", JSON.stringify({
                        appJson: appJson
                    }, metaplexFileReplacer, 2));
                    try {
                        validateApp(appJson);
                        console.info("App JSON valid!");
                    } catch (e) {
                        console.error(e);
                    }
                    return [
                        4,
                        createReleaseJson({
                            releaseDetails: releaseDetails,
                            appDetails: appDetails,
                            publisherDetails: publisherDetails
                        }, signer.publicKey)
                    ];
                case 2:
                    releaseJson = _state.sent();
                    objStringified = JSON.stringify(releaseJson, metaplexFileReplacer, 2);
                    debug("releaseJson=", objStringified);
                    try {
                        validateRelease(JSON.parse(objStringified));
                        console.info("Release JSON valid!");
                    } catch (e) {
                        console.error(e);
                    }
                    return [
                        2
                    ];
            }
        });
    });
    return function validateCommand(_) {
        return _ref.apply(this, arguments);
    };
}();
