// @solana-mobile/dapp-store-cli@0.7.1/lib/upload/CachedStorageDriver.js
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _defineProperties(target, props) {
    for(var i = 0; i < props.length; i++){
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}
function _create_class(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _object_spread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _define_property(target, key, source[key]);
        });
    }
    return target;
}
function _ts_generator(thisArg, body) {
    var f, y, t, g, _ = {
        label: 0,
        sent: function() {
            if (t[0] & 1) throw t[1];
            return t[1];
        },
        trys: [],
        ops: []
    };
    return g = {
        next: verb(0),
        "throw": verb(1),
        "return": verb(2)
    }, typeof Symbol === "function" && (g[Symbol.iterator] = function() {
        return this;
    }), g;
    function verb(n) {
        return function(v) {
            return step([
                n,
                v
            ]);
        };
    }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while(_)try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [
                op[0] & 2,
                t.value
            ];
            switch(op[0]){
                case 0:
                case 1:
                    t = op;
                    break;
                case 4:
                    _.label++;
                    return {
                        value: op[1],
                        done: false
                    };
                case 5:
                    _.label++;
                    y = op[1];
                    op = [
                        0
                    ];
                    continue;
                case 7:
                    op = _.ops.pop();
                    _.trys.pop();
                    continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                        _ = 0;
                        continue;
                    }
                    if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
                        _.label = op[1];
                        break;
                    }
                    if (op[0] === 6 && _.label < t[1]) {
                        _.label = t[1];
                        t = op;
                        break;
                    }
                    if (t && _.label < t[2]) {
                        _.label = t[2];
                        _.ops.push(op);
                        break;
                    }
                    if (t[2]) _.ops.pop();
                    _.trys.pop();
                    continue;
            }
            op = body.call(thisArg, _);
        } catch (e) {
            op = [
                6,
                e
            ];
            y = 0;
        } finally{
            f = t = 0;
        }
        if (op[0] & 5) throw op[1];
        return {
            value: op[0] ? op[1] : void 0,
            done: true
        };
    }
}
import fs from "fs";
import { createHash } from "crypto";
// TODO(jon): We need to manage the removal / replacement of assets in the manifest
export var CachedStorageDriver = /*#__PURE__*/ function() {
    "use strict";
    function CachedStorageDriver(storageDriver, param) {
        var assetManifestPath = param.assetManifestPath;
        _class_call_check(this, CachedStorageDriver);
        _define_property(this, "assetManifest", void 0);
        _define_property(this, "assetManifestPath", void 0);
        _define_property(this, "storageDriver", void 0);
        this.assetManifestPath = assetManifestPath;
        console.info({
            loading: true
        });
        var _this_loadAssetManifest;
        this.assetManifest = (_this_loadAssetManifest = this.loadAssetManifest(assetManifestPath)) !== null && _this_loadAssetManifest !== void 0 ? _this_loadAssetManifest : {
            schema_version: CachedStorageDriver.SCHEMA_VERSION,
            assets: {}
        };
        this.storageDriver = storageDriver;
    }
    _create_class(CachedStorageDriver, [
        {
            key: "getUploadPrice",
            value: function getUploadPrice(bytes) {
                var _this = this;
                return _async_to_generator(function() {
                    return _ts_generator(this, function(_state) {
                        return [
                            2,
                            _this.storageDriver.getUploadPrice(bytes)
                        ];
                    });
                })();
            }
        },
        {
            key: "loadAssetManifest",
            value: function loadAssetManifest(filename) {
                try {
                    return JSON.parse(fs.readFileSync(filename, "utf-8"));
                } catch (error) {
                    console.warn("Failed opening ".concat(filename, "; initializing with a blank asset manifest"));
                    return;
                }
            }
        },
        {
            key: "uploadedAsset",
            value: function uploadedAsset(filename, param) {
                var sha256 = param.sha256;
                var _this_assetManifest_assets_filename;
                if (((_this_assetManifest_assets_filename = this.assetManifest.assets[filename]) === null || _this_assetManifest_assets_filename === void 0 ? void 0 : _this_assetManifest_assets_filename.sha256) === sha256) {
                    return this.assetManifest.assets[filename];
                }
                return null;
            }
        },
        {
            key: "upload",
            value: function upload(file) {
                var _this = this;
                return _async_to_generator(function() {
                    var hash, uploadedAsset, uri;
                    return _ts_generator(this, function(_state) {
                        switch(_state.label){
                            case 0:
                                if (!(file.fileName === "inline.json")) return [
                                    3,
                                    2
                                ];
                                return [
                                    4,
                                    _this.storageDriver.upload(file)
                                ];
                            case 1:
                                return [
                                    2,
                                    _state.sent()
                                ];
                            case 2:
                                hash = createHash("sha256").update(file.buffer).digest("base64");
                                console.info(JSON.stringify({
                                    file: {
                                        name: file.fileName,
                                        disn: file.displayName,
                                        un: file.uniqueName
                                    }
                                }));
                                uploadedAsset = _this.uploadedAsset(file.fileName, {
                                    sha256: hash
                                });
                                if (uploadedAsset) {
                                    console.log("Asset ".concat(file.fileName, " already uploaded at ").concat(uploadedAsset.uri));
                                    return [
                                        2,
                                        uploadedAsset.uri
                                    ];
                                }
                                console.log("Uploading ".concat(file.fileName));
                                return [
                                    4,
                                    _this.storageDriver.upload(file)
                                ];
                            case 3:
                                uri = _state.sent();
                                _this.assetManifest.assets[file.fileName] = {
                                    path: file.fileName,
                                    sha256: hash,
                                    uri: uri
                                };
                                return [
                                    4,
                                    fs.promises.writeFile("".concat(process.cwd(), "/").concat(_this.assetManifestPath), // Something is really weird, I can't seem to stringify `this.assetManifest` straight-up. Here be dragons
                                    JSON.stringify({
                                        assets: _object_spread({}, _this.assetManifest.assets)
                                    }, null, 2), "utf-8")
                                ];
                            case 4:
                                _state.sent();
                                return [
                                    2,
                                    uri
                                ];
                        }
                    });
                })();
            }
        }
    ]);
    return CachedStorageDriver;
}();
// NOTE: this schema version is independent of the publishing JSON schema. It should be updated
// when the AssetManifestSchema or Asset types are updated.
_define_property(CachedStorageDriver, "SCHEMA_VERSION", "0.1");
