// @solana-mobile/dapp-store-publishing-tools@0.7.1/lib/publish/PublishCoreUpdate.js
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _ts_generator(thisArg, body) {
    var f, y, t, g, _ = {
        label: 0,
        sent: function() {
            if (t[0] & 1) throw t[1];
            return t[1];
        },
        trys: [],
        ops: []
    };
    return(g = {
        next: verb(0),
        "throw": verb(1),
        "return": verb(2)
    }, typeof Symbol === "function" && (g[Symbol.iterator] = function() {
        return this;
    }), g);
    function verb(n) {
        return function(v) {
            return step([
                n,
                v
            ]);
        };
    }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while(_)try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [
                op[0] & 2,
                t.value
            ];
            switch(op[0]){
                case 0:
                case 1:
                    t = op;
                    break;
                case 4:
                    _.label++;
                    return {
                        value: op[1],
                        done: false
                    };
                case 5:
                    _.label++;
                    y = op[1];
                    op = [
                        0
                    ];
                    continue;
                case 7:
                    op = _.ops.pop();
                    _.trys.pop();
                    continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                        _ = 0;
                        continue;
                    }
                    if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
                        _.label = op[1];
                        break;
                    }
                    if (op[0] === 6 && _.label < t[1]) {
                        _.label = t[1];
                        t = op;
                        break;
                    }
                    if (t && _.label < t[2]) {
                        _.label = t[2];
                        _.ops.push(op);
                        break;
                    }
                    if (t[2]) _.ops.pop();
                    _.trys.pop();
                    continue;
            }
            op = body.call(thisArg, _);
        } catch (e) {
            op = [
                6,
                e
            ];
            y = 0;
        } finally{
            f = t = 0;
        }
        if (op[0] & 5) throw op[1];
        return {
            value: op[0] ? op[1] : void 0,
            done: true
        };
    }
}
import { createAttestationPayload } from "./PublishCoreAttestation.js";
import { CONTACT_OBJECT_ID, CONTACT_PROPERTY_COMPANY, CONTACT_PROPERTY_EMAIL, CONTACT_PROPERTY_WEBSITE, submitRequestToSolanaDappPublisherPortal, TICKET_OBJECT_ID, TICKET_PROPERTY_ATTESTATION_PAYLOAD, TICKET_PROPERTY_AUTHORIZED_REQUEST, TICKET_PROPERTY_CRITICAL_UPDATE, TICKET_PROPERTY_DAPP_COLLECTION_ACCOUNT_ADDRESS, TICKET_PROPERTY_DAPP_RELEASE_ACCOUNT_ADDRESS, TICKET_PROPERTY_POLICY_COMPLIANT, TICKET_PROPERTY_REQUEST_UNIQUE_ID, TICKET_PROPERTY_TESTING_INSTRUCTIONS, URL_FORM_UPDATE } from "./dapp_publisher_portal.js";
var createUpdateRequest = function() {
    var _ref = _async_to_generator(function(connection, sign, appMintAddress, releaseMintAddress, publisherDetails, solanaMobileDappPublisherPortalDetails, compliesWithSolanaDappStorePolicies, requestorIsAuthorized, criticalUpdate) {
        var _ref, attestationPayload, requestUniqueId, request;
        return _ts_generator(this, function(_state) {
            switch(_state.label){
                case 0:
                    return [
                        4,
                        createAttestationPayload(connection, sign)
                    ];
                case 1:
                    _ref = _state.sent(), attestationPayload = _ref.attestationPayload, requestUniqueId = _ref.requestUniqueId;
                    request = {
                        fields: [
                            {
                                objectTypeId: CONTACT_OBJECT_ID,
                                name: CONTACT_PROPERTY_COMPANY,
                                value: publisherDetails.name
                            },
                            {
                                objectTypeId: CONTACT_OBJECT_ID,
                                name: CONTACT_PROPERTY_EMAIL,
                                value: publisherDetails.email
                            },
                            {
                                objectTypeId: CONTACT_OBJECT_ID,
                                name: CONTACT_PROPERTY_WEBSITE,
                                value: publisherDetails.website
                            },
                            {
                                objectTypeId: TICKET_OBJECT_ID,
                                name: TICKET_PROPERTY_ATTESTATION_PAYLOAD,
                                value: attestationPayload
                            },
                            {
                                objectTypeId: TICKET_OBJECT_ID,
                                name: TICKET_PROPERTY_DAPP_COLLECTION_ACCOUNT_ADDRESS,
                                value: appMintAddress
                            },
                            {
                                objectTypeId: TICKET_OBJECT_ID,
                                name: TICKET_PROPERTY_DAPP_RELEASE_ACCOUNT_ADDRESS,
                                value: releaseMintAddress
                            },
                            {
                                objectTypeId: TICKET_OBJECT_ID,
                                name: TICKET_PROPERTY_REQUEST_UNIQUE_ID,
                                value: requestUniqueId
                            },
                            {
                                objectTypeId: TICKET_OBJECT_ID,
                                name: TICKET_PROPERTY_AUTHORIZED_REQUEST,
                                value: requestorIsAuthorized
                            },
                            {
                                objectTypeId: TICKET_OBJECT_ID,
                                name: TICKET_PROPERTY_POLICY_COMPLIANT,
                                value: compliesWithSolanaDappStorePolicies
                            }
                        ]
                    };
                    if (criticalUpdate) {
                        request.fields.push({
                            objectTypeId: TICKET_OBJECT_ID,
                            name: TICKET_PROPERTY_CRITICAL_UPDATE,
                            value: criticalUpdate
                        });
                    }
                    if (solanaMobileDappPublisherPortalDetails.testing_instructions !== undefined) {
                        request.fields.push({
                            objectTypeId: TICKET_OBJECT_ID,
                            name: TICKET_PROPERTY_TESTING_INSTRUCTIONS,
                            value: solanaMobileDappPublisherPortalDetails.testing_instructions
                        });
                    }
                    return [
                        2,
                        request
                    ];
            }
        });
    });
    return function createUpdateRequest(connection, sign, appMintAddress, releaseMintAddress, publisherDetails, solanaMobileDappPublisherPortalDetails, compliesWithSolanaDappStorePolicies, requestorIsAuthorized, criticalUpdate) {
        return _ref.apply(this, arguments);
    };
}();
export var publishUpdate = function() {
    var _ref = _async_to_generator(function(publishSolanaNetworkInput, param, dryRun) {
        var appMintAddress, releaseMintAddress, publisherDetails, solanaMobileDappPublisherPortalDetails, compliesWithSolanaDappStorePolicies, requestorIsAuthorized, criticalUpdate, updateRequest;
        return _ts_generator(this, function(_state) {
            switch(_state.label){
                case 0:
                    appMintAddress = param.appMintAddress, releaseMintAddress = param.releaseMintAddress, publisherDetails = param.publisherDetails, solanaMobileDappPublisherPortalDetails = param.solanaMobileDappPublisherPortalDetails, compliesWithSolanaDappStorePolicies = param.compliesWithSolanaDappStorePolicies, requestorIsAuthorized = param.requestorIsAuthorized, criticalUpdate = param.criticalUpdate;
                    return [
                        4,
                        createUpdateRequest(publishSolanaNetworkInput.connection, publishSolanaNetworkInput.sign, appMintAddress, releaseMintAddress, publisherDetails, solanaMobileDappPublisherPortalDetails, compliesWithSolanaDappStorePolicies, requestorIsAuthorized, criticalUpdate)
                    ];
                case 1:
                    updateRequest = _state.sent();
                    submitRequestToSolanaDappPublisherPortal(updateRequest, URL_FORM_UPDATE, dryRun);
                    return [
                        2
                    ];
            }
        });
    });
    return function publishUpdate(publishSolanaNetworkInput, _, dryRun) {
        return _ref.apply(this, arguments);
    };
}();
